core/src/main/java/org/lflang/federated/generator/FederateInstance.java

```diff
@@ -698,24 +719,69 @@ public TimeValue findNearestPhysicalActionTrigger(ReactionInstance reaction) {
+                      .add(findNearestPhysicalActionTriggerRecursive(visited, uReaction));
               if (uMinDelay.isEarlierThan(minDelay)) {
                 minDelay = uMinDelay;
               }
             }
           }
         }
-
-      } else if (trigger.getDefinition() instanceof Output) {
-        // Outputs of contained reactions
-        PortInstance outputInstance = (PortInstance) trigger;
-        for (ReactionInstance uReaction : outputInstance.getDependsOnReactions()) {
-          TimeValue uMinDelay = findNearestPhysicalActionTrigger(uReaction);
-          if (uMinDelay.isEarlierThan(minDelay)) {
-            minDelay = uMinDelay;
+      } else if (trigger.getDefinition() instanceof Port) {
+        PortInstance port = (PortInstance) trigger;
+        // Regardless of whether the port is an output or input port of
+        // a contained reactor, recurse on reactions that write to it.
+        for (ReactionInstance uReaction : port.getDependsOnReactions()) {
+          // Ports can form a loop also.
+          if (!visited.contains(uReaction)) {
+            TimeValue uMinDelay = findNearestPhysicalActionTriggerRecursive(visited, uReaction);
+            if (uMinDelay.isEarlierThan(minDelay)) {
+              minDelay = uMinDelay;
+            }
+          }
+        }
+        // The trigger is an input of a contained reactor. If it
+        // another upstream contained reactor connects to it, trace to
+        // the upstream contained reactor's output ports.
+        if (trigger.getDefinition() instanceof Input) {
+          var upstreamPorts = port.eventualSources();
+          for (RuntimeRange<PortInstance> range : upstreamPorts) {
+            PortInstance upstreamPort = range.instance;
+
+            // Get a connection delay value between upstream port and
+            // the current port. This is currently done by finding
+            // the corresponding SendRange from the upstream port
+            // and then extracting a delay value from the connection
+            // contained in the SendRange.
+            // TODO: Find a better way to do this, which likely involves
+            // refactoring SendRange, RuntimeRange.Port, and PortInstance.
+            TimeValue connectionDelay =
+                upstreamPort.getDependentPorts().stream()
+                    .filter(
+                        sRange ->
+                            sRange.destinations.stream()
+                                .map(it -> it.instance)
+                                .anyMatch(port::equals))
+                    .map(sRange -> ASTUtils.getDelay(sRange.connection.getDelay()))
+                    .filter(Objects::nonNull)
+                    .map(TimeValue::fromNanoSeconds)
+                    .findFirst()
+                    .orElse(TimeValue.ZERO);
+
+            for (ReactionInstance uReaction : upstreamPort.getDependsOnReactions()) {
+              if (!visited.contains(uReaction)) {
```

Comment:
Again, I'm not sure this should be skipped just because the upstream has been visited. There may be multiple paths to the upstream.  I suggest replacing `visited` with a set that contains only the "in progress" reactions (those for which we are inside a recursive call to find their delay). Then also cache the values computed for each reaction (just for efficiency, not needed for correctness).